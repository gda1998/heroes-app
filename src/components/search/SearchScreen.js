import { useMemo } from 'react';
import { useNavigate } from 'react-router-dom';
import queryString from 'query-string';
import { useLocation } from 'react-router';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = () => {

    const navigate = useNavigate();
    const { search } = useLocation();

    const { q = '' } = queryString.parse(search);

    const [ form, handleInputChange ] = useForm({
        'searchText' : q
    });

    const { searchText } = form;

    const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

    const handleSubmit = (e) => {
        e.preventDefault();
        navigate(`?q=${ searchText }`);
    }

    return (
        <>
            <h1>Search Screen</h1>
            <hr />
            <div className="row">
                <div className="col-md-5">
                    <h4>Search Form</h4>
                    <hr />
                    <form onSubmit={ handleSubmit } className="input-group">
                        <input
                            type="text"
                            name="searchText"
                            placeholder="Find your hero"
                            className="form-control"
                            autoComplete="off"
                            value={ searchText }
                            onChange={ handleInputChange }
                        />
                        <button 
                            type="submit" 
                            className="btn btn-outline-primary">
                            Button
                        </button>
                    </form>
                </div>
                <div className="col-md-7">
                    <h4>Results</h4>
                    <hr />
                    {
                        (q === '') &&
                        <div className="alert alert-info">
                            Search a hero
                        </div>
                    }
                    {
                        (q !== '' && heroesFiltered.length === 0) &&
                        <div className="alert alert-warning">
                            The hero {q} was not found
                        </div>
                    }
                    {
                        heroesFiltered.map( hero => (
                            <HeroCard key={hero.id} {...hero} />
                        ))
                    }
                </div>
            </div>
        </>
    )
}
