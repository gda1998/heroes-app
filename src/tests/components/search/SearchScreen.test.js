import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('Pruebas en <SearchSreen />', () => {

    const historyMock = {
        push: jest.fn()
    };

    const getWrapper = (path) => {
        return mount(
            <MemoryRouter initialEntries={[ path ]}>
                <Route 
                    path="/search" 
                    component={ () => <SearchScreen history={ historyMock } /> } 
                />
            </MemoryRouter>
        );
    }

    beforeEach( () => jest.clearAllMocks() );

    test('El componente se debe de mostrar correctamente', () => {
        const wrapper = getWrapper('/search');
        const alertInfoElement = wrapper.find('.alert-info').text().trim();

        expect(wrapper).toMatchSnapshot();
        expect(alertInfoElement).toBe('Search a hero');
    });

    test('Debe de mostrar a Batman y el input con el valor del queryString', () => {
        const wrapper = getWrapper('/search?q=batman');
        const inputValue = wrapper.find('input').prop('value');
        expect(wrapper).toMatchSnapshot();
        expect(inputValue).toBe('batman');
    });

    test('Debe de mostrar un error si no se encuentra el heroe', () => {
        const wrapper = getWrapper('/search?q=batman123');
        const alertElement = wrapper.find('.alert-warning');
        expect(alertElement.exists()).toBe(true);
    });
    
    test('Debe de llamar el push del history', () => {
        const querySeachValue = 'batman';
        const wrapper = getWrapper(`/search?q=${ querySeachValue }`);
        wrapper.find('input').simulate('change', {
            target: { name: 'searchText', value: querySeachValue }
        });
        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        expect(historyMock.push).toHaveBeenCalledWith(`?q=${ querySeachValue }`)
    });
    
});
