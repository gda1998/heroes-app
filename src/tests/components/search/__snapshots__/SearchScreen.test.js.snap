// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchSreen /> Debe de mostrar a Batman y el input con el valor del queryString 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-md-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        className="input-group"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary"
          type="submit"
        >
          Button
        </button>
      </form>
    </div>
    <div
      className="col-md-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="col"
      >
        <div
          className="card mb-3"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row g-0"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Batman"
                className="card-img"
                loading="lazy"
                src="./assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  MÃ¡s...
                </a>
              </div>
            </div>
          </div>
        </div>
         
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en <SearchSreen /> El componente se debe de mostrar correctamente 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-md-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        className="input-group"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary"
          type="submit"
        >
          Button
        </button>
      </form>
    </div>
    <div
      className="col-md-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>,
]
`;
